### PR TITLE
Fix running on musl libc

### PR DESCRIPTION
**Pull request by bodgit**
This is following on from the work in PR #522 which reached the point that uwsgi compiled on musl libc (Alpine Linux) however it doesn't run because it fails on `pthread_mutexattr_setprotocol(&attr, PTHREAD_PRIO_INHERIT)` which returns `ENOTSUP` and the code treated any non-zero return value as fatal.

This PR handles this specific case while still considering any other non-zero return value as an error the same as before.

With this change uwsgi now runs on Alpine Linux.

### CODE REVIEW

File: core/lock.c [hunk 1]
```diff
@@ -74,6 +74,9 @@ int pthread_mutexattr_setprotocol (pthread_mutexattr_t *__attr,
 struct uwsgi_lock_item *uwsgi_lock_fast_init(char *id) {
 
 	pthread_mutexattr_t attr;
+#ifdef EOWNERDEAD
```
Comment (xrmx) [inline core/lock.c]: can't you move it inside if (uwsgi_pthread_robust_mutexes_enabled) block instead?

File: core/lock.c [hunk 2]
```diff
@@ -95,9 +98,18 @@ struct uwsgi_lock_item *uwsgi_lock_fast_init(char *id) {
 #define PTHREAD_MUTEX_ROBUST PTHREAD_MUTEX_ROBUST_NP
 #endif
 	if (uwsgi_pthread_robust_mutexes_enabled) {
-		if (pthread_mutexattr_setprotocol(&attr, PTHREAD_PRIO_INHERIT)) {
-			uwsgi_log("unable to set PTHREAD_PRIO_INHERIT\n");
-			exit(1);
+		if ((ret = pthread_mutexattr_setprotocol(&attr, PTHREAD_PRIO_INHERIT)) != 0) {
```
Comment (xrmx) [inline core/lock.c]: Would an if looks more natural here? I don't have an answer, just asking :)

**Comment by xrmx** [comment]
Can you please add all the description of the issue you added here to the commit message too? It's the commit message the source of truth, not github. Thanks a lot!

File: core/lock.c (hunk 1 above)
Comment (bodgit) [inline core/lock.c]: Sure. Personally I don't like declaring variables all over the place but I suppose it makes sense here.

**Comment by bodgit** [comment]
Done

**Comment by patrick91** [comment]
@bodgit I was trying to compile your version of uwsgi on a docker machine based on alpine, but I get this error:

```
    *** uWSGI linking ***
    gcc -o /usr/bin/uwsgi -L/usr/lib -Wl,-rpath,/usr/lib core/utils.o core/protocol.o core/socket.o core/logging.o core/master.o core/master_utils.o core/emperor.o core/notify.o core/mule.o core/subscription.o core/stats.o core/sendfile.o core/async.o core/master_checks.o core/fifo.o core/offload.o core/io.o core/static.o core/websockets.o core/spooler.o core/snmp.o core/exceptions.o core/config.o core/setup_utils.o core/clock.o core/init.o core/buffer.o core/reader.o core/writer.o core/alarm.o core/cron.o core/hooks.o core/plugins.o core/lock.o core/cache.o core/daemons.o core/errors.o core/hash.o core/master_events.o core/chunked.o core/queue.o core/event.o core/signal.o core/strings.o core/progress.o core/timebomb.o core/ini.o core/fsmon.o core/mount.o core/metrics.o core/plugins_builder.o core/sharedarea.o core/fork_server.o core/webdav.o core/zeus.o core/rpc.o core/gateway.o core/loop.o core/cookie.o core/querystring.o core/rb_timers.o core/transformations.o core/uwsgi.o proto/base.o proto/uwsgi.o proto/http.o proto/fastcgi.o proto/scgi.o proto/puwsgi.o lib/linux_ns.o core/yaml.o core/dot_h.o core/config_py.o plugins/python/python_plugin.o plugins/python/pyutils.o plugins/python/pyloader.o plugins/python/wsgi_handlers.o plugins/python/wsgi_headers.o plugins/python/wsgi_subhandler.o plugins/python/web3_subhandler.o plugins/python/pump_subhandler.o plugins/python/gil.o plugins/python/uwsgi_pymodule.o plugins/python/profiler.o plugins/python/symimporter.o plugins/python/tracebacker.o plugins/python/raw.o plugins/gevent/gevent.o plugins/gevent/hooks.o plugins/ping/ping_plugin.o plugins/cache/cache.o plugins/nagios/nagios.o plugins/rrdtool/rrdtool.o plugins/carbon/carbon.o plugins/rpc/rpc_plugin.o plugins/corerouter/cr_common.o plugins/corerouter/cr_map.o plugins/corerouter/corerouter.o plugins/fastrouter/fastrouter.o plugins/http/http.o plugins/http/keepalive.o plugins/http/https.o plugins/http/spdy3.o plugins/ugreen/ugreen.o plugins/signal/signal_plugin.o plugins/syslog/syslog_plugin.o plugins/rsyslog/rsyslog_plugin.o plugins/logsocket/logsocket_plugin.o plugins/router_uwsgi/router_uwsgi.o plugins/router_redirect/router_redirect.o plugins/router_basicauth/router_basicauth.o plugins/zergpool/zergpool.o plugins/redislog/redislog_plugin.o plugins/mongodblog/mongodblog_plugin.o plugins/router_rewrite/router_rewrite.o plugins/router_http/router_http.o plugins/logfile/logfile.o plugins/router_cache/router_cache.o plugins/rawrouter/rawrouter.o plugins/router_static/router_static.o plugins/sslrouter/sslrouter.o plugins/spooler/spooler_plugin.o plugins/cheaper_busyness/cheaper_busyness.o plugins/symcall/symcall_plugin.o plugins/transformation_tofile/tofile.o plugins/transformation_gzip/gzip.o plugins/transformation_chunked/chunked.o plugins/transformation_offload/offload.o plugins/router_memcached/router_memcached.o plugins/router_redis/router_redis.o plugins/router_hash/router_hash.o plugins/router_expires/expires.o plugins/router_metrics/plugin.o plugins/transformation_template/tt.o plugins/stats_pusher_socket/plugin.o -lpthread -lm -rdynamic -ldl -ldl -lm -lpython2.7 -lcryptplugins/ugreen/ugreen.o: In function `u_green_init':
    ugreen.c:(.text+0xca): undefined reference to `getcontext'
    plugins/ugreen/ugreen.o: In function `u_green_schedule_to_req':
    ugreen.c:(.text+0x2db): undefined reference to `swapcontext'
    ugreen.c:(.text+0x329): undefined reference to `makecontext'
    plugins/ugreen/ugreen.o: In function `u_green_schedule_to_main':
    ugreen.c:(.text+0x39a): undefined reference to `swapcontext'
    collect2: error: ld returned 1 exit status

    *** error linking uWSGI ***

    ----------------------------------------
Command "/usr/bin/python -c "import setuptools, tokenize;__file__='/src/uwsgi/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" develop --no-deps" failed with error code 1 in /src/uwsgi/
The command '/bin/sh -c pip install -e git+git://github.com/bodgit/uwsgi.git@3dff394af24e9a51ab56569b16497f7d6cc68ab8#egg=uwsgi' returned a non-zero code: 1
```

Here is the docker file:

```
FROM alpine:3.3

RUN apk add --no-cache python python-dev gcc musl-dev linux-headers
RUN apk add --no-cache --virtual=build-dependencies wget ca-certificates && \
    wget "https://bootstrap.pypa.io/get-pip.py" -O /dev/stdout | python


RUN apk add --no-cache git
# RUN pip install uwsgi
RUN pip install -e git+git://github.com/bodgit/uwsgi.git@3dff394af24e9a51ab56569b16497f7d6cc68ab8#egg=uwsgi
```
 
Am I missing something?

**Comment by bodgit** [comment]
@patrick91 The `ugreen` plugin doesn't build. There's discussion around the lack of `*context()` in musl libc in #522.

Building uwsgi using a profile that doesn't include ugreen should work, either by doing `UWSGI_PROFILE=<core or some other profile> pip install ...` or one of the other methods mentioned in the documentation.

I'd like to fix this issue too as it prevents doing a simple `pip install` but I'm not sure how one would go about fixing it.

**Comment by patrick91** [comment]
I confirm that with `UWSGI_PROFILE=core` it works :)

**Comment by xrmx** [comment]
@patrick91 care to file another issue (workaround included) so we can track it? Given that musl will be receiving more exposure that could be an hot topic for which the earlier we find a solution the better.

**Comment by xrmx** [comment]
Merged, thanks @bodgit!

**Pull request merged**